txpool/txpool.go

```diff
@@ -64,126 +76,122 @@ type signer interface {
 //
-// There are fundamentally 2 queues in the txpool module:
-// - Account based transactions (accountQueues)
-// - Global valid transactions, from any account (pendingQueue)
+// When consensus needs to process promoted transactions,
+// the pool generates a queue of "executable" transactions. These
+// transactions are the first-in-line of some promoted queue,
+// ready to be written to the state (primaries).
 type TxPool struct {
 	logger     hclog.Logger
 	signer     signer
 	forks      chain.ForksInTime
 	store      store
 	idlePeriod time.Duration
 
-	// Unsorted min heap of transactions per account.
-	// The heap is min nonce based
-	accountQueuesLock sync.Mutex
-	accountQueues     map[types.Address]*accountQueueWrapper
+	// map of all accounts registered by the pool
+	accounts accountsMap
 
-	// Max price heap for all transactions that are valid
-	pendingQueue *txPriceHeap
+	// all the primaries sorted by max gas price
+	executables *pricedQueue
 
-	// Min price heap for all remote transactions
-	remoteTxns *txPriceHeap
+	// lookup map keeping track of all
+	// transactions present in the pool
+	index lookupMap
 
-	// Lookup map that keeps track of txns present in the pool
-	txnLookupMap map[types.Hash]*types.Transaction
-
-	// Lock for the txn lookup map
-	txnLookupMapLock sync.RWMutex
+	// networking stack
+	topic *network.Topic
 
-	// Gauge for measuring pool capacity
+	// gauge for measuring pool capacity
 	gauge slotGauge
 
-	// Networking stack
-	topic *network.Topic
+	// channels on which the pool's event loop
+	// does dispatching/handling requests.
+	enqueueReqCh chan enqueueRequest
+	promoteReqCh chan promoteRequest
 
-	// Flag indicating if the current node is a sealer,
+	// shutdown channel
+	shutdownCh chan struct{}
+
+	// flag indicating if the current node is a sealer,
 	// and should therefore gossip transactions
 	sealing bool
 
-	// Flag indicating if the current node is running in dev mode
+	// flag indicating if the current node is running in dev mode (used for testing)
 	dev bool
 
-	// Whether local transaction handling should be disabled
-	noLocals bool
-
-	// Addresses that should be treated as local
-	locals *localAccounts
-
-	// priceLimit is a lower threshold for gas price
-	priceLimit uint64
-
-	// Notification channel used so signal added transactions to the pool
-	NotifyCh chan struct{}
+	// prometheus API
+	metrics *Metrics
 
-	// Indicates which txpool operator commands should be implemented
+	// indicates which txpool operator commands should be implemented
 	proto.UnimplementedTxnPoolOperatorServer
-
-	metrics *Metrics
 }
 
-// NewTxPool creates a new pool for transactions
+// Returns a new pool for processing incoming transactions.
 func NewTxPool(
 	logger hclog.Logger,
-	sealing bool,
-	locals []types.Address,
-	noLocals bool,
-	priceLimit uint64,
-	maxSlots uint64,
 	forks chain.ForksInTime,
 	store store,
 	grpcServer *grpc.Server,
 	network *network.Server,
 	metrics *Metrics,
+	config *Config,
```

Comment:
Can we move out this as a single param (struct)?

Reply:
I suggest a separate PR for this.

Reply:
@dbrajovic let's create an issue for that 👍